core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -81,9 +86,17 @@ class ZooKeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
       val responseQueue = new ArrayBlockingQueue[Req#Response](requests.size)
 
       requests.foreach { request =>
-        send(request) { response =>
-          responseQueue.add(response)
-          countDownLatch.countDown()
+        inFlightRequests.acquire()
+        try {
+          send(request) { response =>
+            responseQueue.add(response)
+            inFlightRequests.release()
+            countDownLatch.countDown()
+          }
+        } catch {
+          case e: Throwable =>
```

Comment:
Hmm, will a ZK operation with callback ever throw an exception?

Reply:
If you're talking about an exception being thrown not from the callback but from the user thread calling:
- `zooKeeper.exists`
- `zooKeeper.getData`
- `zooKeeper.getChildren`
- `zooKeeper.create`
- `zooKeeper.setData`
- `zooKeeper.delete`
- `zooKeeper.getACL`
- `zooKeeper.setACL`

then the answer is yes. I described how in an earlier comment here:
https://github.com/apache/kafka/pull/3860#discussion_r148075028

> More specifically, org.apache.zookeeper.ZooKeeper async apis internally call org.apache.zookeeper.common.PathUtils.validatePath, which throws IllegalArgumentException if the path is invalid based on some local validation checks.